plugins/discovery-azure-classic/src/test/java/org/elasticsearch/cloud/azure/classic/AbstractAzureComputeServiceTestCase.java

```diff
@@ -60,13 +81,107 @@ protected Settings nodeSettings(int nodeOrdinal) {
+                for (Map.Entry<String, DiscoveryNode> node : nodes.entrySet()) {
+                    final String name = node.getKey();
+                    final DiscoveryNode discoveryNode = node.getValue();
+
+                    RoleInstance instance = new RoleInstance();
+                    instance.setInstanceName(name);
+                    instance.setHostName(discoveryNode.getHostName());
+                    instance.setPowerState(RoleInstancePowerState.Started);
+
+                    // Set the private IP address
+                    final TransportAddress transportAddress = discoveryNode.getAddress();
+                    instance.setIPAddress(transportAddress.address().getAddress());
+
+                    // Set the public IP address
+                    final InstanceEndpoint endpoint = new InstanceEndpoint();
+                    endpoint.setName(Discovery.ENDPOINT_NAME_SETTING.getDefault(Settings.EMPTY));
+                    endpoint.setVirtualIPAddress(transportAddress.address().getAddress());
+                    endpoint.setPort(transportAddress.address().getPort());
+                    instance.setInstanceEndpoints(new ArrayList<>(Collections.singletonList(endpoint)));
+                    instances.add(instance);
+                }
+
+                final HostedServiceGetDetailedResponse.Deployment deployment = new HostedServiceGetDetailedResponse.Deployment();
+                deployment.setName("dummy");
+                deployment.setDeploymentSlot(DeploymentSlot.Production);
+                deployment.setStatus(DeploymentStatus.Running);
+                deployment.setRoleInstances(new ArrayList<>(Collections.unmodifiableList(instances)));
+
+                final HostedServiceGetDetailedResponse response = new HostedServiceGetDetailedResponse();
+                response.setDeployments(newSingletonArrayList(deployment));
+
+                return response;
+            };
+        }
+
+        /**
+         * Defines a {@link AzureUnicastHostsProvider} for testing purpose that is able to resolve
+         * network addresses for Azure instances running on the same host but different ports.
+         */
+        @Override
+        protected AzureUnicastHostsProvider createUnicastHostsProvider(final Settings settings,
+                                                                       final AzureComputeService azureComputeService,
+                                                                       final TransportService transportService,
+                                                                       final NetworkService networkService) {
+            return new AzureUnicastHostsProvider(settings, azureComputeService, transportService, networkService) {
+                @Override
+                protected String resolveInstanceAddress(final HostType hostType, final RoleInstance instance) {
+                    if (hostType == HostType.PRIVATE_IP) {
```

Comment:
why is this needed? we set the private IP on the RoleInstance, no?

Reply:
For private IP, AzureUnicastHostProvider resolves the network address from the RoleInstance.getIPAddress() which returns an IP address like 127.0.0.1 with no port. I suppose that this is OK on Azure, but in tests we can have multiple nodes on the same IP address. 

This is why it needs a slightly different way to resolve private IP addresses in order to include the port number. Then it will be correctly translated by [transportService.addressesFromString()](https://github.com/elastic/elasticsearch/blob/5a7c8bb04ed135be1823a4db6b08f7481d7d22c0/plugins/discovery-azure-classic/src/main/java/org/elasticsearch/discovery/azure/classic/AzureUnicastHostsProvider.java#L244) that only ping 1 port.

Before this change the test used ports 9400 and 9401 which were corresponding to the default ports used in TcpTransport.